### PR TITLE
Add validation for duplicate item names

### PR DESCRIPTION
**Pull request by sudhirverma**
Fix: #549

### CODE REVIEW

**Comment by sudhirverma** [comment]
testPR

File: src/openshift/application.ts [hunk 1]
```diff
@@ -24,17 +24,24 @@ export class Application extends OpenShiftItem {
...
+    static async getApplicationName(project) {
         return await vscode.window.showInputBox({
             prompt: "Application name",
-            validateInput: (value: string) => {
-                const validationMessage = Application.emptyName('Empty application name', value.trim());
-                if (!validationMessage) return Application.validateMatches('Not a valid Application name. Please use lower case alphanumeric characters or "-", and must start and end with an alphanumeric character', value);
+            validateInput: async (value: string) => {
+                let validationMessage = Application.emptyName('Empty application name', value.trim());
+                if (!validationMessage) validationMessage = Application.validateMatches('Not a valid Application name. Please use lower case alphanumeric characters or "-", and must start and end with an alphanumeric character', value);
+                if (!validationMessage) validationMessage = Application.lengthName('Storage name is to long', value);
```
Comment (jrichter1) [inline src/openshift/application.ts]: You mean 'Application name is too long'?

File: src/openshift/openshiftItem.ts [hunk 2]
```diff
@@ -32,10 +32,23 @@ export abstract class OpenShiftItem {
         return new Promise((res) => setTimeout(res, timeout));
     }
 
+    static openshiftData(data, value) {
```
Comment (jrichter1) [inline src/openshift/openshiftItem.ts]: Why name a method that checks for name uniqueness 'openshiftData'? That makes zero sense.
Also, the type definitions are missing.

File: src/openshift/project.ts [hunk 3]
```diff
@@ -28,6 +20,24 @@ export class Project extends OpenShiftItem {
             .catch((error) => Promise.reject(`Failed to create Project with error '${error}'`));
     }
 
+    private static async getProjectName() {
+        return await vscode.window.showInputBox({
+            prompt: "Mention Project name",
+            validateInput: async (value: string) => {
+                let validationMessage = Project.emptyName('Empty Project name', value.trim());
+                if (!validationMessage) validationMessage = Project.validateMatches('Not a valid Project name. Please use lower case alphanumeric characters or "-", and must start and end with an alphanumeric character', value);
+                if (!validationMessage) validationMessage = Project.lengthName('Project name is to long', value);
```
Comment (jrichter1) [inline src/openshift/project.ts]: should be 'too long'

File: src/openshift/service.ts [hunk 4]
```diff
@@ -25,31 +25,35 @@ export class Service extends OpenShiftItem {
...
     }
 
+    private static async getServiceName(application: OpenShiftObject, serviceTemplateName: string) {
+        return await vscode.window.showInputBox({
+            value: serviceTemplateName,
+            prompt: "Service Name",
+            validateInput: async (value: string) => {
+                let validationMessage = Service.emptyName('Empty Service name', value.trim());
+                if (!validationMessage) validationMessage = Service.validateMatches('Not a valid Service name. Please use lower case alphanumeric characters or "-", and must start and end with an alphanumeric character', value);
+                if (!validationMessage) validationMessage = Service.lengthName('Service name is to long', value);
```
Comment (jrichter1) [inline src/openshift/service.ts]: again, 'too long'

File: src/openshift/storage.ts [hunk 5]
```diff
@@ -42,6 +39,23 @@ export class Storage extends OpenShiftItem {
         }
     }
 
+    private static async getStorageName(component: OpenShiftObject) {
+        return await vscode.window.showInputBox({
+            prompt: "Specify the Storage name",
+            validateInput: async (value: string) => {
+                let validationMessage = Storage.emptyName('Empty Storage name', value.trim());
+                if (!validationMessage) validationMessage = Storage.validateMatches('Not a valid Storage name. Please use lower case alphanumeric characters or "-", and must start and end with an alphanumeric character', value);
+                if (!validationMessage) validationMessage = Storage.lengthName('Storage name is to long', value);
```
Comment (jrichter1) [inline src/openshift/storage.ts]: 'too long' again

File: src/openshift/application.ts (hunk 1 above)
Comment (sudhirverma) [reply src/openshift/application.ts]: Yes thanks @jrichter1

**Comment by sudhirverma** [comment]
@jrichter1 please review this PR

File: src/openshift/application.ts [hunk 6]
```diff
@@ -24,12 +24,15 @@ export class Application extends OpenShiftItem {
         return null;
     }
 
-    static async getApplicationName() {
+    static async getApplicationName(project: OpenShiftObject) {
+        const applicationList: Array<OpenShiftObject> = await OpenShiftItem.odo.getApplications(project);
         return await vscode.window.showInputBox({
             prompt: "Application name",
```
Comment (jrichter1) [inline src/openshift/application.ts]: I just realized the prompt message has a different template for each item. We've got 
 - "Application name"
 - "Mention Project name"
 - "Service Name"
 - "Specify the Storage name"

How about we choose one and use it everywhere, for consistency's sake.
 Thing is, I don't particularly like either of these. Personally, I'd be happier to see something in lines of:
 - "Choose a name for the $ITEM"

@mohitsuman @dgolovin feel free to chime in on what you'd like best :)

File: src/openshift/openshiftItem.ts [hunk 7]
```diff
@@ -32,16 +32,29 @@ export abstract class OpenShiftItem {
         return new Promise((res) => setTimeout(res, timeout));
     }
 
+    static validateName(data: Array<OpenShiftObject>, value: string) {
```
Comment (jrichter1) [inline src/openshift/openshiftItem.ts]: Can you at least name this to be indicative of what it checks for? 'validateName' is a little too broad for what it does. Something in line with the other validators could do.

File: src/openshift/component.ts [hunk 8]
```diff
@@ -201,6 +200,7 @@ export class Component extends OpenShiftItem {
             validateInput: async (value: string) => {
                 let validationMessage = Component.emptyName('Empty Component name', value.trim());
                 if (!validationMessage) validationMessage = Component.validateMatches('Not a valid Component name. Please use lower case alphanumeric characters or "-", and must start and end with an alphanumeric character', value);
+                if (!validationMessage) validationMessage = Component.lengthName('Component name is to long', value);
```
Comment (jrichter1) [inline src/openshift/component.ts]: too is still missing an o

File: src/openshift/application.ts (hunk 6 above)
Comment (dgolovin) [reply src/openshift/application.ts]: @jrichter1 @mohitsuman I would stick to "Application name" with description "Please provide an Application name" the same pattern is used in `Git: Create Branch...`.

**Comment by sudhirverma** [comment]
@jrichter1  please review this PR

File: src/openshift/openshiftItem.ts [hunk 9]
```diff
@@ -32,16 +32,42 @@ export abstract class OpenShiftItem {
         return new Promise((res) => setTimeout(res, timeout));
     }
 
+    static validateUniqueName(data: Array<OpenShiftObject>, value: string) {
+        const openshiftObject =  data.find((openshiftObject) =>  openshiftObject.getName() === value);
+        return openshiftObject && `This name is already used, please enter different name.`;
+    }
+
+    static async getName(message: string, data: Array<OpenShiftObject>) {
```
Comment (jrichter1) [inline src/openshift/openshiftItem.ts]: please ensure the methods have defined return types

File: src/openshift/openshiftItem.ts [hunk 10]
```diff
@@ -32,16 +32,42 @@ export abstract class OpenShiftItem {
...
+        return await window.showInputBox({
+            prompt: `Please provide an ${message}`,
```
Comment (jrichter1) [inline src/openshift/openshiftItem.ts]: Looks good, however, the sentence becomes grammatically incorrect once ${message} starts with a consonant (i.e. 'provide an component').

File: test/openshift/application.test.ts [hunk 11]
```diff
@@ -68,7 +69,43 @@ suite('Openshift/Application', () => {
 
                 await Application.create(projectItem);
 
-                expect(result).is.equals('Empty application name');
+                expect(result).is.equals('Empty Application name');
```
Comment (jrichter1) [inline test/openshift/application.test.ts]: It makes no functional difference (this is just for the sake of grammar), but there is no need to combine is and equals :)
The point of the fluent API verbs here is to essentially create sentences for readability and "expect something is equals" is not really a "valid" sentence. Chai recommends using "expect something to equal", shorthand for which is "expect something equals".

File: src/openshift/openshiftItem.ts (hunk 9 above)
Comment (sudhirverma) [reply src/openshift/openshiftItem.ts]: done

Comment (jrichter1) [reply src/openshift/openshiftItem.ts]: just string will do, undefined is a subtype of all other types anyway

Comment (sudhirverma) [reply src/openshift/openshiftItem.ts]: done

**Comment by sudhirverma** [comment]
@jrichter1 can you this PR now I have done the changes

**Comment by dgolovin** [comment]
Rebased on top of master.

File: src/openshift/openshiftItem.ts [hunk 12]
```diff
@@ -32,16 +32,42 @@ export abstract class OpenShiftItem {
...
+        return openshiftObject && `This name is already used, please enter different name.`;
+    }
+
+    static async getName(message: string, data: Array<OpenShiftObject>): Promise<string> {
+        return await window.showInputBox({
+            prompt: `Provide ${message}`,
+            validateInput: (value: string) => {
+                let validationMessage = OpenShiftItem.emptyName(`Empty ${message}`, value.trim());
+                if (!validationMessage) validationMessage = OpenShiftItem.validateMatches(`Not a valid ${message}. Please use lower case alphanumeric characters or "-", and must start and end with an alphanumeric character`, value);
+                if (!validationMessage) validationMessage = OpenShiftItem.lengthName(`${message} is too long`, value);
```
Comment (dgolovin) [inline src/openshift/openshiftItem.ts]: It should work for projects, applications ans services, but not components. The reason is a component name entered by user then combined with application name when actual deployment config created.
so getName function should have additional parameter than would offset 63 by length of  '${application-name}-}, default value would be 0.

File: src/openshift/openshiftItem.ts [hunk 13]
```diff
@@ -32,16 +32,42 @@ export abstract class OpenShiftItem {
...
+        });
+    }
+
     static emptyName(message: string, value: string) {
-        return validator.isEmpty(value) ? message : undefined;
+        return validator.isEmpty(value) ? message : null;
+    }
+
+    static lengthName(message: string, value: string) {
+        return validator.isLength(value, 0, 63) ? null : message;
```
Comment (dgolovin) [inline src/openshift/openshiftItem.ts]: Here we would have 

```
63 - offset
```

**Comment by sudhirverma** [comment]
@jrichter1 and @dgolovin review this pr

**Comment by jrichter1** [comment]
I'm satisfied

**Pull request merged**